### PR TITLE
Introduce the Encoding component

### PR DESCRIPTION
**Pull request by veewee**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | 

#### Summary

This component introduces `xml_encode()` and `xml_decode()` so that you can use XML with the ease of JSON.

Examples:

```php
$data = xml_decode(
    file_get_contents('file.xml'),
    validator(xsd_validator('some-schema.xsd'))
);

// You can read or change the data
$data['root']['name'] = 'MyName';

// And finally convert it back to XML!
$xml = xml_encode($data, utf8(), pretty_print());
```

```php
$data = typed(
    <<<EOXML
        <root>
           <item>
               <id>1</id>
               <name>X</name>
               <category>A</category>
               <category>B</category>
               <category>C</category>
           </item>     
        </root>
    EOXML,
    shape([
        'root' => shape([
            'item' => shape([
                'id' => int(),
                'name' => string(),
                'category' => vector(string()),
            ])
        ])
    ]),
    validator(xsd_validator('some-schema.xsd'))
);
```

### CODE REVIEW

File: src/Xml/Dom/Builder/escaped_value.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?php
```
Comment (veewee) [inline src/Xml/Dom/Builder/escaped_value.php:1]: TODO : doc

File: src/Xml/Dom/Locator/Namespaces/linked_namespaces.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<?php
```
Comment (veewee) [inline src/Xml/Dom/Locator/Namespaces/linked_namespaces.php:1]: TODO : doc

File: src/Xml/Dom/Locator/Namespaces/recursive_linked_namespaces.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?php
```
Comment (veewee) [inline src/Xml/Dom/Locator/Namespaces/recursive_linked_namespaces.php:1]: TODO : doc

File: src/Xml/Dom/Locator/document_element.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?php
```
Comment (veewee) [inline src/Xml/Dom/Locator/document_element.php:1]: TODO : doc

File: src/Xml/Encoding/encode.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+use VeeWee\Xml\Dom\Document;
+use VeeWee\Xml\Encoding\Exception\EncodingException;
+use VeeWee\Xml\Exception\RuntimeException;
+use function VeeWee\Xml\Encoding\Internal\Encoder\Builder\normalize_data;
+use function VeeWee\Xml\Encoding\Internal\Encoder\Builder\root;
+
+/**
+ * @param list<callable(DOMDocument): DOMDocument> $configurators
+ *
+ * @throws RuntimeException
```
Comment (veewee) [inline src/Xml/Encoding/encode.php:19]: TODO : wrap all exceptions in EncodingException

File: src/Xml/Encoding/decode.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+use VeeWee\Xml\Dom\Document;
+use VeeWee\Xml\Encoding\Internal\Decoder\Context;
+use VeeWee\Xml\Exception\RuntimeException;
+use function VeeWee\Xml\Dom\Locator\document_element;
+use function VeeWee\Xml\Encoding\Internal\Decoder\Builder\element;
+
+/**
+ * @param list<callable(DOMDocument): DOMDocument> $configurators
+ *
+ * @throws RuntimeException
```
Comment (veewee) [inline src/Xml/Encoding/decode.php:17]: TODO : wrap all exceptions in EncodingException

File: docs/encoding.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+By using these functions there is no need for ever using the `SimpleXML` library again!
+
+## Example
+
+```php
+use function VeeWee\Xml\Dom\Configurator\pretty_print;
+use function VeeWee\Xml\Dom\Configurator\utf8;
+use function VeeWee\Xml\Dom\Configurator\validator;
+use function VeeWee\Xml\Dom\Validator\xsd_validator;
+use function VeeWee\Xml\Encoding\decode as xml_decode;
```
Comment (bortefi) [inline docs/encoding.md:13]: I assume it will mostly be used as `xml_decode` could make the functions default name.

Comment (veewee) [reply docs/encoding.md]: Yes, changed it

**Pull request merged**